File: abc/341/c.rb
# TLE
H, W, N = gets.split.map(&:to_i)
T = gets.chomp
S = H.times.map { gets.chomp }

ans = 0
H.times do |i|
  W.times do |j|
    next if S[i][j] == '#'
    now_i = i
    now_j = j
    flag = true
    N.times do |k|
      if T[k] == 'D' && now_i + 1 < H && S[now_i + 1][now_j] == '.'
        now_i += 1
      elsif T[k] == 'U' && now_i - 1 >= 0 && S[now_i - 1][now_j] == '.'
        now_i -= 1
      elsif T[k] == 'R' && now_j + 1 < W && S[now_i][now_j + 1] == '.'
        now_j += 1
      elsif T[k] == 'L' && now_j - 1 >= 0 && S[now_i][now_j - 1] == '.'
        now_j -= 1
      else
        flag = false
        break
      end
    end
    ans += 1 if flag
  end
end

puts ans

File: abc/341/c.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int main() {
  int H, W, N;
  cin >> H >> W >> N;

  string T;
  cin >> T;

  vector<string> S(H);
  for(int i = 0; i < H; i++) {
    cin >> S[i];
  }

  int ans = 0;
  for(int i=0; i<H; i++) {
    for(int j=0; j<W; j++) {
      if(S[i][j] == '#') continue;

      int now_i = i;
      int now_j = j;
      bool flag = true;

      for(int k=0; k<N; k++) {
        if(T[k] == 'D' && now_i + 1 < H && S[now_i + 1][now_j] == '.') {
          now_i += 1;
        } else if(T[k] == 'U' && now_i - 1 >= 0 && S[now_i - 1][now_j] == '.') {
          now_i -= 1;
        } else if(T[k] == 'R' && now_j + 1 < W && S[now_i][now_j + 1] == '.') {
          now_j += 1;
        } else if(T[k] == 'L' && now_j - 1 >= 0 && S[now_i][now_j - 1] == '.') {
          now_j -= 1;
        } else {
          flag = false;
          break;
        }
      }
        if(flag) ans += 1;
    }
  }

  cout << ans << endl;
}
